### ISSUE TITLE
Inactive colors behaving weirdly

### ISSUE THREAD

**Issue by filipekiss** [fenetikm/falcon#16]
Hey @fenetikm, thank you so much for Falcon! I love the colors.

I've tried to move my setup to use Falcon all around (NeoVim + iTerm2 + TMUX) and something seems odd and I hope you can point me to a direction fo what I am doing wrong.

When I move to a pane that does not have NeoVim open, the color doesn't seem to change for the inactive one:

![image](https://user-images.githubusercontent.com/48519/44496175-b2f4a100-a649-11e8-8ef0-d2781c7d3d7b.png)

I've checked to see if the `autocmd` for the color change were working and they are. So I ran `hi InactiveWindow guibg=yellow` and then this happens:

![image](https://user-images.githubusercontent.com/48519/44496365-8e4cf900-a64a-11e8-8245-71537dd404b8.png)

 And when I move focus out of NeoVim to another pane:

![image](https://user-images.githubusercontent.com/48519/44496380-9e64d880-a64a-11e8-8a28-81a928365b9c.png)
(See how the last focused window is _not_ yellow?)

So, two things: Is there a way to make the InactiveWindow bg be applied to the whole window/buffer and not just to the size of the file?

And second: Shouldn't all NeoVim buffers be marked as "inactive" when I'm on another Tmux pane?

Thank you so much for your help!

**Comment by fenetikm** [comment]
Yeah I really need to add in a bit more detail on how to set this up as it's a bit fiddly.

So to get the whole inactive / active thing going you need:

- NeoVim (not supported in Vim)
- 24-bit color working inside of tmux, inside of vim
- the latest version of falcon which makes this stuff configurable: https://github.com/fenetikm/falcon/blob/b9b186b5aa5ae96267ccf3e35b8a34efee4ba0e8/plugin/falcon.vim#L22
- in your tmux config somewhere:
```
set -g window-style "bg=#151521"
set -g window-active-style "bg=#020221"
```
- in your NeoVim config:
```
"24bit color
set termguicolors

let g:falcon_lightline = 1
let g:lightline.colorscheme='falcon'
let g:falcon_background = 0
let g:falcon_inactive = 1

colorscheme falcon
```
(the lightline stuff is optional)

Let me know if you have all of the above and how you go.

**Comment by filipekiss** [comment]
So I updated my NeoVim (built from source, latest commit) and checked for truecolor:
![image](https://user-images.githubusercontent.com/48519/44497844-dfacb680-a651-11e8-9914-7ef18c651cc4.png)
Everything seems to be ok. (Left is Terminal inside NeoVim and right is pure shell)

`echo exists('+winhighlight')` returns 1 and `g:falcon_inactive` is equal 1 (and `g:falcon_background` is 0);

I've imported the falcon theme into iTerm as mentioned in the Wiki. The Tmux Panes seem ok.

**Comment by fenetikm** [comment]
Oh, that was the other thing, need to be on a recent version of NeoVim (note to self).

**Comment by filipekiss** [comment]
I'm guessing my colors might be a bit off. Comparing my screenshot to the one on the README, I've noticed the current active tmux window should be orange and mine is grey (on TMUX, that is):

| Mine | README |
| --- | --- |
| ![image](https://user-images.githubusercontent.com/48519/44498273-e0dee300-a653-11e8-86d7-b86248039796.png) | <img width="161" alt="image" src="https://user-images.githubusercontent.com/48519/44498301-023fcf00-a654-11e8-887b-cf3c73741f52.png"> |

**Comment by fenetikm** [comment]
hah, sorry, screenshot is slightly behind - I have gone back to grey.

**Comment by fenetikm** [comment]
in my tmux conf I currently have this:
```
source-file ~/Documents/Work/internal/vim/colors/falcon/tmux/falcon.tmux.conf

# the following is falcon colour specific
set -g status-left "#[fg=white] #S #[fg=#57575e]│ "
set -g status-right "#(~/.config/tmux/itunes_current_track.zsh) #[fg=#57575e]│ #(~/.config/tmux/tw_status.sh) #[fg=#57575e]│ #[fg=white]%d/%m/%Y %H:%M "
set -g status-right-length 200
```

**Comment by fenetikm** [comment]
So I am importing the falcon one then overwriting the status a bit since I have a script to show the itunes track and a taskwarrior script.

**Comment by filipekiss** [comment]
Yeah, I'm starting to guess it's more a NeoVim issue than a Falcon one.

Can you confirm something for me? Open a Tmux window, split it in two and open an empty Buffer on NeoVim and then focus the other pane. Does NeoVim change the bg color to the Inactive one?

**Comment by fenetikm** [comment]
Yeah, it works perfectly for me. 

The key to the way it works is that NeoVim should _only_ set the background for inactive buffers and leave the background unset otherwise, thus taking it's colour from Tmux which will be either the Tmux inactive colour or the Tmux active colour.

It used to be that there was no background set anywhere in falcon, however, I noted that had some issues in certain places so now, the background is set for `Normal` and `NonText` but is then _unset_ via the `plugin/falcon.vim` file which looks at the settings.

So the tests I would run, would be:
- open up NeoVim just in iTerm2, no Tmux - does the active inactive thing work?
- open up three iTerm2 instances, check that the background colour in NeoVim, Tmux inside of Neovim and iTerm2 all match
- open up Tmux inside of iTerm2, make some splits, check that the colours change as expected
- open up NeoVim inside of Tmux (no extra splits), check that the colours changes as expected
- open up NeoVim inside of Tmux, two splits (one NeoVim, one shell), create two splits in NeoVim, check that moving between NeoVim splits and Tmux works as expected re active / inactive

**Comment by fenetikm** [comment]
I remember I had some interesting issues at one stage due to iTerm2 background being wrong (I have sinced moved away from iTerm2 for performance reasons).

**Comment by fenetikm** [comment]
Actually I have this in my config:
```
"turn off the cursorline when not in a window
autocmd InsertLeave,WinEnter * set cursorline
autocmd InsertEnter,WinLeave * set nocursorline
```

**Comment by fenetikm** [comment]
maybe that is forcing an extra redraw?

**Comment by fenetikm** [comment]
And also this for reloading files, probably not relevant:
```
" Reload when gaining focus
autocmd FocusGained,BufEnter * :silent! !
```

**Comment by filipekiss** [comment]
So, here's the deal:

Opened `nvim` outside tmux, just iTerm and split in 3 (Active Pane is bottom-right):

![image](https://user-images.githubusercontent.com/48519/44498829-4da7ac80-a657-11e8-9492-f7eac3afc09a.png)

Then, I manually ran `hi ActiveWindow guibg=NONE | hi InactiveWindow guibg=#151521` and it becomes like this (Active Pane is bottom-right):

![image](https://user-images.githubusercontent.com/48519/44498853-7760d380-a657-11e8-8a71-cd343b16e94c.png)

Better, but if the bottom-right pane is the inactive one, only the first line stays `inactive` (Active Pane is top):

![image](https://user-images.githubusercontent.com/48519/44498869-98292900-a657-11e8-87e8-d01babf02562.png)

Same thing if I ran this inside tmux

**Comment by filipekiss** [comment]
Also I should note that I copied the autocmds you suggested before trying this, so they're definetely not the responsibles for doing this

**Comment by fenetikm** [comment]
can you possibly post a link to your nvim config?

**Comment by fenetikm** [comment]
it looks like to me that the falcon plugin (vs the colorscheme) isn't doing it's thing

**Comment by fenetikm** [comment]
essentially this code:
```vim
" NeoVim has support for changing background colour depending on active or not
if !exists('g:falcon_inactive')
  let g:falcon_inactive = 0
endif

if exists('+winhighlight') && g:falcon_inactive == 1
  hi ActiveWindow guibg=NONE | hi InactiveWindow guibg=#151521
  set winhighlight=Normal:ActiveWindow,NormalNC:InactiveWindow
endif

" Put in a background colour for gui
if !exists('g:falcon_background')
  let g:falcon_background = 1
endif

if !has("gui_running") || g:falcon_background == 0
  hi NonText guifg=#36363a ctermfg=237 guibg=NONE ctermbg=NONE gui=NONE cterm=NONE
  hi Normal guifg=#b4b4b9 ctermfg=249 guibg=NONE ctermbg=NONE gui=NONE cterm=NONE
endif
```

**Comment by filipekiss** [comment]
My [vim config](https://github.com/filipekiss/vimfiles) is kinda huge. I've disabled all plugins to see if something was conflicting but still no luck. I'm loading it via vim-plug. I'll try to run the plugin code manually and see if it works, hang on.

**Comment by fenetikm** [comment]
thanks

**Comment by fenetikm** [comment]
I am using plug, but I am just including it ala:
```
Plug '~/Documents/Work/internal/vim/colors/falcon'
```

**Comment by fenetikm** [comment]
Great - progress!

If you look in the falcon repo, you will see that there is `plugin/falcon.vim` - my suggestion would be to check if that is being called and when. Also what the values are of the global variables, especially if it is hitting that line that you have just manually added in. Maybe pepper the code with `echom` statements or what have you.

Would be great if you can post back here and so I can write up a separate wiki page on getting this all working.

**Comment by filipekiss** [comment]
So, a few things I've picked up (and I wasn't really able to determine the issue yet):

- If I do not load my `plugin/color.vim` file, the whole inactive thing works as expected
- If I load the file mentioned above, the "dimming" of the background color stops working
- Weirdly enough, what makes the whole thing stops working is `colorscheme falcon`

I'll try to run the colorscheme command somewhere else and see if it's the loading order that's causing this.

Ran `:scriptnames` and the `plugin/falcon.vim` is being loaded way before the `color.vim` and the color scheme itself

**Comment by fenetikm** [comment]
hrm, what is currently in your `color.vim` file?

I just made a minimal  `.vimrc`:
```
set nocompatible " disable compatibility with vi
syntax enable "enable syntax highlighting

" Assumes vim-plug is installed.
call plug#begin()
Plug 'fenetikm/falcon'
call plug#end()

"enable 24bit support.
set termguicolors

" enables the active / inactive functionality.
let g:falcon_background = 0
let g:falcon_inactive = 1

colorscheme falcon
```
... and everything works fine. It's sounding like something clobbering the falcon plugin code... and so expected behaviour of sorts.

**Comment by filipekiss** [comment]
My `color.vim` has only this:

```
if has('termguicolors')
    set termguicolors
end

try
    colorscheme falcon
catch
    colorscheme ron
endtry
```

So here's the thing: In tmux it works fine but outside tmux it does not. My understanding is that TMUX is able to set the inactive pane BG to a certain color (and when the vim pane has it's BG set to NONE you see that color) whereas iTerm isn't able to do that. I really don't care about this since I never work outside TMUX but I'd be happy to continue debugging if you think it's worth.

**Comment by fenetikm** [comment]
Oh when you say it doesn't work outside of Tmux in iTerm - are you referring to iTerm splits? If that's the case then yeah, that isn't supported AFAIK.

I'll try modifying my falcon plugin to use an autocmd approach like you have above and then hopefully we can call it day ;)

**Comment by filipekiss** [comment]
No, I actually mean vim splits:

![image](https://user-images.githubusercontent.com/48519/44658546-9b197600-a9d7-11e8-9b3c-709eca155bbc.png)

I'm guessing the commandsare depending on it being loaded on a specific order. I've tried the vim syntax test and the groups for ActiveWindow and InactiveWindow  were not being set properly withouth the `autocmd`.

If you wish I can open a pull-request with the `autocmd` patch :)

Thank you so much for your patience!

**Comment by fenetikm** [comment]
Ok, cool, thanks for the clarification.

All good re the PR - no need to bother.

I want to see it with my own eyes first ;)

**Comment by fenetikm** [comment]
Ok, `develop` branch now has the autocmd version. Also added in your guard statement which is always a good idea. I removed FocusGained - felt that might be a bit much.

Let me know what you think.

**Comment by filipekiss** [comment]
`develop` is hella fine! Working perfectly! Thank you. 🙏
